### PR TITLE
Transaction Service Refactor to use a protocol and broker pattern

### PR DESCRIPTION
**Pull request by philipr-za**
## Description
This PR refactors the 3 most complex long running processes in the Transaction service. The Sending protocol, Broadcast protocol and the Chain monitoring protocol.

These protocols all have multiple complex asynchronous stages and they were all handled in the main service code making them difficult to trace and maintain. The refactor introduces the idea of a “protocol” which is a struct and the logic to execute a long running task that requires complex asynchronous input and processing. The protocol is spawned as an async task and to receive input when it is created it registers channels with the Service task along which the required data will be sent when it arrives at the service.

The main service task still acts as a broker where all input streams (Event stream, comms stack message subscriptions, Service API calls etc) are read and when such an event/message is received if a Protocol has registered interest in the event it will be sent along the channel to the Protocol task.

This refactor makes the writing, reading and maintenance of a complex asynchronous protocol much easier as it can all be written as a single flow of logic. Giving a protocol its own task is also much neater in terms of managing where it awaits.

## Motivation and Context
The async task handling in the Tx service had gotten to complex to easily maintain and update.

## How Has This Been Tested?
Tests have been updated

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+{
+    id: u64,
+    resources: TransactionServiceResources<TBackend>,
+    timeout: Duration,
+    base_node_public_key: CommsPublicKey,
+    mempool_response_receiver: Option<Receiver<MempoolServiceResponse>>,
+    base_node_response_receiver: Option<Receiver<BaseNodeProto::BaseNodeServiceResponse>>,
+}
+
+impl<TBackend> TransactionBroadcastProtocol<TBackend>
```
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs:68]: 👀 here

File: base_layer/wallet/src/transaction_service/protocols/transaction_chain_monitoring_protocol.rs [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+};
+use tari_crypto::tari_utilities::{hex::Hex, Hashable};
+use tari_p2p::tari_message::TariMessageType;
+use tokio::time::delay_for;
+
+const LOG_TARGET: &str = "wallet::transaction_service::protocols::chain_monitoring_protocol";
+
+/// This protocol defines the process of monitoring a mempool and base node to detect when a Broadcast transaction is
+/// Mined or leaves the mempool in which case it should be cancelled
+pub struct TransactionChainMonitoringProtocol<TBackend>
```
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/protocols/transaction_chain_monitoring_protocol.rs:60]: 👀 here

File: base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    tari_amount::MicroTari,
+    transaction::{KernelFeatures, TransactionError},
+    transaction_protocol::{proto, recipient::RecipientSignedMessage},
+    SenderTransactionProtocol,
+};
+use tari_p2p::tari_message::TariMessageType;
+
+const LOG_TARGET: &str = "wallet::transaction_service::protocols::send_protocol";
+
+pub struct TransactionSendProtocol<TBackend>
```
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs:53]: 👀 here

File: base_layer/wallet/src/transaction_service/service.rs [hunk 4]
```diff
@@ -244,27 +238,26 @@ where
...
+            JoinHandle<Result<u64, TransactionServiceProtocolError>>,
         > = FuturesUnordered::new();
 
-        let mut broadcast_timeout_futures: FuturesUnordered<BoxFuture<'static, TxId>> = FuturesUnordered::new();
-        let mut mined_request_timeout_futures: FuturesUnordered<BoxFuture<'static, TxId>> = FuturesUnordered::new();
+        let mut transaction_chain_monitoring_protocol_handles: FuturesUnordered<
+            JoinHandle<Result<u64, TransactionServiceProtocolError>>,
+        > = FuturesUnordered::new();
 
         loop {
```
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/service.rs:254]: Maybe some 👀 here

File: base_layer/wallet/src/transaction_service/protocols/transaction_receive_protocol.rs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// pub struct TransactionReceiveProtocol {
```
Comment (sdbondi) [inline base_layer/wallet/src/transaction_service/protocols/transaction_receive_protocol.rs:23]: TODO: refactor the transaction receiver code :)

File: base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+                base_node_response = base_node_response_receiver.select_next_some() => {
+                    received_base_node_response = Some(base_node_response);
+                },
+                () = delay => {
+                },
+            }
+
+            // Handle a received Mempool Response
+            if let Some(response) = received_mempool_response {
+                if self.handle_mempool_response(response).await? {
```
Comment (sdbondi) [inline base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs:206]: Why not have this inside the `select!` branch above?

File: base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+                    break;
+                }
+            }
+
+            // Handle a received Base node response
+            if let Some(response) = received_base_node_response {
+                if self.handle_base_node_response(response).await? {
```
Comment (sdbondi) [inline base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs:213]: Same here

File: base_layer/wallet/src/transaction_service/mod.rs [hunk 8]
```diff
@@ -153,9 +153,9 @@ where T: TransactionBackend + Clone + 'static
         let mempool_response_stream = self.mempool_response_stream();
         let base_node_response_stream = self.base_node_response_stream();
 
-        let (publisher, subscriber) = bounded(100);
+        let (publisher, _) = channel(200);
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply base_layer/wallet/src/transaction_service/mod.rs]: sure

File: base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+                    let _ = self.resources.event_publisher.send(Arc::new(
+                        TransactionEvent::TransactionSendDiscoveryComplete(tx_id, false),
+                    ));
+                    error!(target: LOG_TARGET, "Transaction Send message failed to send");
+                    return Err(TransactionServiceProtocolError::new(
+                        self.id,
+                        TransactionServiceError::OutboundSendFailure,
+                    ));
+                },
+                SendMessageResponse::PendingDiscovery(response_channel) => match response_channel.await {
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs]: That does help

File: base_layer/wallet/src/transaction_service/protocols/transaction_receive_protocol.rs (hunk 5 above)
Comment (philipr-za) [reply base_layer/wallet/src/transaction_service/protocols/transaction_receive_protocol.rs:23]: For a later PR

File: base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs (hunk 6 above)
Comment (philipr-za) [reply base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs]: artifact of a previous implementation, done.

File: base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+                    completed_tx.transaction.clone().into(),
+                )),
+            };
+
+            self.resources
+                .outbound_message_service
+                .send_direct(
+                    self.base_node_public_key.clone(),
+                    OutboundEncryption::None,
+                    OutboundDomainMessage::new(TariMessageType::MempoolRequest, mempool_request.clone()),
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs:158]: you need it

File: base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs (hunk 7 above)
Comment (philipr-za) [reply base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs]: you need it

**Pull request merged**